File: Programmers/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

bool isValid(const string& str) {
    stack<char> st;

    for (char c : str) 
    {
        if (c == '(' || c == '[' || c == '{') 
        {
            st.push(c);
        } 
        else
        {
            if (st.empty()) return false;
            
            char top = st.top();
            st.pop();
            
            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) return false;
        }
    }
    return st.empty();
}

int solution(string s) {
    int answer = 0;
    int n = s.size();

    for (int i = 0; i < n; ++i)
    {
        string rotated = s.substr(i) + s.substr(0, i);
        
        if (isValid(rotated)) answer++;
    }
    
    return answer;
}

File: Programmers/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        int n = s.length();
        
        for (int i = 0; i < n; i++) {
            String rotated = s.substring(i) + s.substring(0, i);
            
            if (isValid(rotated)) answer++;
        }
        
        return answer;
    }
    
    private boolean isValid(String str) {
        Stack<Character> stack = new Stack<>();

        for (char c : str.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;
                
                char top = stack.pop();
                
                if ((c == ')' && top != '(') ||
                    (c == ']' && top != '[') ||
                    (c == '}' && top != '{')) return false;
            }
        }
        
        return stack.isEmpty();
    }
}